modules/core/04-channel/keeper/multihop_packet_test.go

```diff
@@ -38,38 +46,273 @@ func (suite *MultihopTestSuite) TestRecvPacket() {
+
+			suite.SetupConnections()
+
+			suite.A().ChannelID = ibctesting.FirstChannelID
+			suite.Z().ChannelID = ibctesting.FirstChannelID
+
+			// manually creating channel prevents next recv sequence from being set
+			channel := types.NewChannel(types.OPEN, types.ORDERED, types.NewCounterparty(suite.A().ChannelConfig.PortID, suite.A().ChannelID), suite.Z().GetConnectionHops(), suite.Z().ChannelConfig.Version)
+			suite.Z().Chain.App.GetIBCKeeper().ChannelKeeper.SetChannel(suite.Z().Chain.GetContext(), suite.Z().ChannelConfig.PortID, suite.Z().ChannelID, channel)
+
+			*packet = types.NewPacket(ibctesting.MockPacketData, 1, suite.A().ChannelConfig.PortID, suite.A().ChannelID, suite.Z().ChannelConfig.PortID, suite.Z().ChannelID, defaultTimeoutHeight, disabledTimeoutTimestamp)
+
+			// manually set packet commitment
+			suite.A().Chain.App.GetIBCKeeper().ChannelKeeper.SetPacketCommitment(suite.A().Chain.GetContext(), suite.A().ChannelConfig.PortID, suite.A().ChannelID, packet.GetSequence(), types.CommitPacket(suite.A().Chain.App.AppCodec(), packet))
+			packetHeight = suite.A().Chain.LastHeader.GetHeight()
+
+			suite.Z().Chain.CreateChannelCapability(suite.Z().Chain.GetSimApp().ScopedIBCMockKeeper, suite.Z().ChannelConfig.PortID, suite.Z().ChannelID)
+			channelCap = suite.Z().Chain.GetChannelCapability(suite.Z().ChannelConfig.PortID, suite.Z().ChannelID)
+
+			err = suite.A().UpdateClient()
+			suite.Require().NoError(err)
+			err = suite.Z().UpdateClient()
+			suite.Require().NoError(err)
+		}, false},
+		{"receipt already stored", false, func() {
+			expError = types.ErrNoOpMsg
+			suite.SetupChannels()
+
+			packet, packetHeight, err = suite.A().
+				SendPacket(defaultTimeoutHeight, disabledTimeoutTimestamp, ibctesting.MockPacketData)
+			suite.Require().NoError(err)
+
+			sequence := packet.GetSequence()
+			suite.Z().Chain.App.GetIBCKeeper().ChannelKeeper.SetPacketReceipt(suite.Z().Chain.GetContext(), suite.Z().ChannelConfig.PortID, suite.Z().ChannelID, sequence)
+			channelCap = suite.Z().Chain.GetChannelCapability(suite.Z().ChannelConfig.PortID, suite.Z().ChannelID)
+		}, false},
+		{"validation failed", false, func() {
+			// skip error code check, downstream error code is used from light-client implementations
+
+			// packet commitment not set resulting in invalid proof
+			suite.SetupChannels()
+
+			*packet = types.NewPacket(ibctesting.MockPacketData, 1, suite.A().ChannelConfig.PortID, suite.A().ChannelID, suite.Z().ChannelConfig.PortID, suite.Z().ChannelID, defaultTimeoutHeight, disabledTimeoutTimestamp)
+			packetHeight = suite.A().Chain.LastHeader.GetHeight()
+
+			channelCap = suite.Z().Chain.GetChannelCapability(suite.Z().ChannelConfig.PortID, suite.Z().ChannelID)
+		}, false},
 	}
 
+	packet = &types.Packet{}
+
 	for _, tc := range testCases {
 		tc := tc
 		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
 			suite.SetupTest() // reset
+			expError = nil    // must explicitly set for failed cases
+			dstProofHeight = nil
```

Comment:
After my latest patch, this will not be needed. Instead QueryPacketProof will return a proofHeight which can be used in place of `suite.Z().ProofHeight()` for example.

Reply:
Done